### PR TITLE
CB: Impls for ConcatMaps and Arbitrary

### PR DESCRIPTION
**Pull request by drostron**
Addresses: #1600 and #1601

### CODE REVIEW

File: couchbase/src/main/scala/quasar/physical/couchbase/planner/MapFuncPlanner.scala [hunk 1]
```diff
@@ -219,9 +219,29 @@ final class MapFuncPlanner[F[_]: Monad: NameGenerator, T[_[_]]: Recursive: ShowT
         partialQueryString(n1qlStr)
       )
     case MakeMap(a1, a2)                          =>
-      partialQueryString(s"""object_add({}, ${n1ql(a1)}, ${n1ql(a2)})""").point[M]
+      val a1N1ql  = n1ql(a1)
+      val a2N1ql  = n1ql(a2)
+      val n1qlStr = s"""object_add({}, ${n1ql(a1)}, ${n1ql(a2)})"""
```
Comment (mossprescott) [inline couchbase/src/main/scala/quasar/physical/couchbase/planner/MapFuncPlanner.scala]: Should be `s"""object_add({}, $a1n1ql, $a2N1ql)`, or else remove the `val`s.

File: couchbase/src/main/scala/quasar/physical/couchbase/planner/MapFuncPlanner.scala [hunk 2]
```diff
@@ -219,9 +219,29 @@ final class MapFuncPlanner[F[_]: Monad: NameGenerator, T[_[_]]: Recursive: ShowT
...
+           |  a2:   $a2N1ql
+           |  n1ql: $n1qlStr""".stripMargin('|')
+      ))).as(
+        partialQueryString(n1qlStr)
+      )
     case ConcatArrays(a1, a2)                     =>
-      partialQueryString(s"array_concat(${n1ql(a1)}, ${n1ql(a2)})").point[M]
+      val a1N1ql  = n1ql(a1)
+      val a2N1ql  = n1ql(a2)
+      val n1qlStr = s"array_concat(${n1ql(a1)}, ${n1ql(a2)})"
```
Comment (mossprescott) [inline couchbase/src/main/scala/quasar/physical/couchbase/planner/MapFuncPlanner.scala]: Similar. I guess you need the `val`s for logging.

File: it/src/main/resources/tests/boulderLikePop.test [hunk 3]
```diff
@@ -13,6 +12,8 @@
 
   "predicate": "containsExactly",
 
+  "ignoreFieldOrder": true,
```
Comment (mossprescott) [inline it/src/main/resources/tests/boulderLikePop.test]: Why ignore the field order here? This looks correct; I think we only want to use that flag in cases where we know MongoDB gets it wrong.

**Comment by mossprescott** [comment]
Needs changelog, by the way

**Comment by mossprescott** [comment]
Not really an issue for this PR, but the way PhaseResults are being produced here isn't useful, I don’t think (without being able to run the planner I haven’t actually looked at the result).

If you run in the REPL with `debug = 2`, each PhaseResult should be a verbose but comprehensible summary of one stage of compilation. It looks like you're producing one result per node. I guess that's ok if it’s useful for debugging a the moment, but it's something we'll want to straighten out at some point. Have you actually been using the log?

**Comment by jdegoes** [comment]
@drostron The other thing is we really need to check on mount or mount access that the version of Couchbase used is sufficient for the features we're using. Probably not an immediate priority since Couchbase will probably generate some error (assuming we propagate it).

**Comment by drostron** [comment]
@mossprescott I've been using PhaseResults frequently for debugging at the repl. What's there was a remedy for ad-hoc printlns. Moving towards N1QL generation as a single PhaseResult sounds good, #1656.

**Comment by drostron** [comment]
@jdegoes sounds good. Once #1621 is in place we'll propagate errors from Couchbase more consistently. I also created #1657.

File: couchbase/src/main/scala/quasar/physical/couchbase/planner/MapFuncPlanner.scala (hunk 2 above)
Comment (drostron) [inline couchbase/src/main/scala/quasar/physical/couchbase/planner/MapFuncPlanner.scala]: Right, just missed updating the n1qlStrs after PhaseResult additions.

File: it/src/main/resources/tests/boulderLikePop.test (hunk 3 above)
Comment (drostron) [inline it/src/main/resources/tests/boulderLikePop.test]: Up until now, I was assuming that N1QL offered the ability to specify result set field order but a further look finds this is not the case. e.g.

```
cbq> select v.z, v.y from (select value { "z": 1, "y": 2 }) v;
{
    ⋮
    "results": [
        {
            "y": 2,
            "z": 1
        }
    ],
    ⋮
}
```

**Comment by drostron** [comment]
I was hoping a test using `zips.data` would be possible on Travis, but it is timing out consistently. That test is pretty tailored to `zips.data` so I'll switch it out for another.

**Comment by drostron** [comment]
Ready for another look. We need to decide how to handle N1QL not supporting result set field ordering.

**Comment by drostron** [comment]
I propose we find a solution to the field order concern in another PR via #1668.

File: couchbase/src/test/scala/quasar/physical/couchbase/BasicQueryEnablementSpec.scala [hunk 4]
```diff
@@ -94,17 +94,13 @@ class BasicQueryEnablementSpec
...
       "select name from `beer-sample` limit 1",
-      """select value v from (select value _4 from (select value _2[0:_1[0]] from (select value (select value {"name": _5.["name"]} from (select value ifmissing(v.`value`, v) from `beer-sample` v) as _5) from (select value (select value [])) as _0) as _2 let _1 = (select value 1 let _6 = (select value []))) as _3 unnest _3 _4) as v""")
+      """select value v from (select value _4 from (select value _2[0:least(array_length(_2), _1[0])] from (select value (select value {"name": _5.["name"]} from (select value ifmissing(v.`value`, v) from `beer-sample` v) as _5) from (select value (select value [])) as _0) as _2 let _1 = (select value 1 let _6 = (select value []))) as _3 unnest _3 _4) as v""")
 
     testSql2ToN1qlPending(
       "select name from `beer-sample` order by name",
       pending("#1545"))
 
     testSql2ToN1ql(
-      """select address from `beer-sample` where name = "Brasserie de Silly"""",
```
Comment (drostron) [inline couchbase/src/test/scala/quasar/physical/couchbase/BasicQueryEnablementSpec.scala]: The tests in this file are fragile and intended for removal as they are covered by query integration tests. Currently it's too easy for the generated N1QL strings to change, still be valid and correct, yet trigger a test failure. There is however opportunity and value for unit style tests against a richer N1QL representation when it is available.

Comment (drostron) [inline couchbase/src/test/scala/quasar/physical/couchbase/BasicQueryEnablementSpec.scala]: Agreed, and looking forward to it. The current representation was a path towards getting everything initially stood-up and we'll have increased motivation for the richer representation as we require things like N1QL specific transformations or finer grained tests.

**Comment by alissapajer** [comment]
LGTM will merge when jenkins gives green.

**Pull request merged**